diff --git a/src/client.c b/src/client.c
index 464d78d..d938750 100644
--- a/src/client.c
+++ b/src/client.c
@@ -49,10 +49,12 @@ PgDatabase *prepare_auth_database(PgSocket *client)
 	PgDatabase *auth_db = NULL;
 	const char *auth_dbname = client->db->auth_dbname ? client->db->auth_dbname : cf_auth_dbname;
 
-	if (!auth_dbname)
-		return client->db;
+	if (!auth_dbname) {
+		auth_db = client->db;
+	} else {
+		auth_db = find_database(auth_dbname);
+	}
 
-	auth_db = find_database(auth_dbname);
 	if (!auth_db) {
 		slog_error(client, "authentication database \"%s\" is not configured.", auth_dbname);
 		disconnect_client(client, true, "bouncer config error");
@@ -68,6 +70,12 @@ PgDatabase *prepare_auth_database(PgSocket *client)
 		return NULL;
 	}
 
+	if (auth_db->admin) {
+		slog_error(client, "cannot use the reserved \"%s\" database as an auth_dbname", auth_db->dbname);
+		disconnect_client(client, true, "bouncer config error");
+		return NULL;
+	}
+
 	return auth_db;
 }
 
@@ -798,7 +806,7 @@ static bool handle_client_startup(PgSocket *client, PktHdr *pkt)
 			return false;
 		}
 
-		if (client->pool && !client->wait_for_user_conn && !client->wait_for_user) {
+		if (client->pool && !client->wait_for_user_conn && !client->wait_for_user && !global_writer) {
 			disconnect_client(client, true, "client re-sent startup pkt");
 			return false;
 		}
